okio/jvm/src/main/java/okio/BufferedSource.kt

```diff
@@ -572,6 +572,27 @@ interface BufferedSource : Source, ReadableByteChannel {
   @Throws(IOException::class)
   fun rangeEquals(offset: Long, bytes: ByteString, bytesOffset: Int, byteCount: Int): Boolean
 
+  /**
+   * Returns a new `BufferedSource` that can read data from this `BufferedSource` without consuming
+   * it. The returned source becomes invalid once this source is next read or closed.
+   *
+   * For example, we can use `peek()` to lookahead and read the same data multiple times.
+   *
+   * ```
+   * val buffer = Buffer()
+   * buffer.writeUtf8("abcdefghi")
+   *
+   * buffer.readUtf8(3) // returns "abc", buffer contains "defghi"
+   *
+   * val peek = buffer.peek()
+   * peek.readUtf8(3) // returns "def", buffer contains "defghi"
+   * peek.readUtf8(3) // returns "ghi", buffer contains "defghi"
+   *
+   * buffer.readUtf8(3) // returns "def", buffer contains "ghi"
+   * ```
+   */
+  fun peek(): BufferedSource
```

Comment:
Since buffering is so easy, maybe this should return `Source` for a better understanding of the invalid-state behavior?

Reply:
I have a hard time seeing anyone doing anything but buffering the source as soon as they get it.